Add `as` to cast wrapped protocols to a concrete type (#819)
swiftlang/swift-java: Sources/SwiftJavaDocumentation/Documentation.docc/SupportedFeatures.md

(`any (A & B)`) are not currently supported; such requirements are simply omitted from
the generated box, and composite returns are not extracted.

#### Downcasting to the concrete type (`as`)

> Note: Downcasting returned protocol values is currently only supported in JNI mode.

A value returned as `any P` / `some P` preserves its concrete dynamic type, so it can be

> recovered — the equivalent of Swift's `as?`. Every imported protocol `interface` therefore

Reviewer: ```suggestion
recovered; This is equivalent to Swift's `as?`, of a checked `instanceof` cast in Java. Every imported protocol `interface` therefore
```

exposes an `as` method:

```java
<T extends JNISwiftInstance> Optional<T> as(Class<T> type, SwiftArena arena);
<T extends JNISwiftInstance> Optional<T> as(Class<T> type); // uses the default arena

> ```

Reviewer: I'd prefer to not have signatures copy pasted into docs

> `type` must be a concrete jextracted type (a `class`, `struct`, or `enum` binding — the

Reviewer: ```suggestion
`type` must be a concrete jextracted type.
```

Reviewer: Too claude-ish long sentences explaining the impl

> `JNISwiftInstance` bound enforces this at compile time). The cast succeeds only when the

Reviewer: ```suggestion
The cast succeeds only when the
```

value's dynamic Swift type is exactly that type, in which case you receive a fresh binding
registered in the given arena; otherwise the result is `Optional.empty()`.

Continuing the `Greeter` example, where `makeEnglishGreeter` returns `any Greeter` backed
by a concrete `EnglishGreeter`:

```java
try (var arena = SwiftArena.ofConfined()) {
    Greeter greeter = MySwiftLibrary.makeEnglishGreeter("World", arena);

>     // Recover the concrete type — its `struct`-only members become available:

Reviewer: ```suggestion
```

    Optional<EnglishGreeter> english = greeter.as(EnglishGreeter.class, arena);
    assertEquals("World", english.orElseThrow().getName());

    // A cast to the wrong dynamic type yields an empty Optional:
    assertTrue(greeter.as(DanishGreeter.class, arena).isEmpty());
}
```

> The cast returns `Optional.empty()` — rather than throwing — whenever it cannot apply: when

Reviewer: ```suggestion
The cast returns an empty optional if the cast fails, which might happen when
```

the dynamic type differs, when `type` is not a concrete jextracted type (e.g. a generic type

> or another protocol), or when the receiver is not backed by a wrapped Swift value.

Reviewer: ```suggestion
or another protocol).
```

Isn't the past point the same? if it's not jextracted, it's not backed by swift.
